src/realm/object-store/sync/sync_session.cpp

```diff
@@ -865,47 +865,51 @@ void SyncSession::handle_progress_update(uint64_t downloaded, uint64_t downloada
     m_progress_notifier.update(downloaded, downloadable, uploaded, uploadable, download_version, snapshot_version);
 }
 
-static sync::Session::Config::ClientReset make_client_reset_config(RealmConfig session_config, DBRef&& fresh_copy,
-                                                                   bool recovery_is_allowed)
+static sync::Session::Config::ClientReset make_client_reset_config(const RealmConfig& base_config,
+                                                                   const std::shared_ptr<SyncConfig>& sync_config,
+                                                                   DBRef&& fresh_copy, bool recovery_is_allowed)
 {
-    REALM_ASSERT(session_config.sync_config->client_resync_mode != ClientResyncMode::Manual);
+    REALM_ASSERT(sync_config->client_resync_mode != ClientResyncMode::Manual);
 
     sync::Session::Config::ClientReset config;
-    config.mode = session_config.sync_config->client_resync_mode;
+    config.mode = sync_config->client_resync_mode;
     config.fresh_copy = std::move(fresh_copy);
     config.recovery_is_allowed = recovery_is_allowed;
 
-    session_config.sync_config = std::make_shared<SyncConfig>(*session_config.sync_config); // deep copy
-    session_config.scheduler = nullptr;
-    if (session_config.sync_config->notify_after_client_reset) {
-        config.notify_after_client_reset = [config = session_config](VersionID previous_version, bool did_recover) {
-            auto local_coordinator = RealmCoordinator::get_coordinator(config);
-            REALM_ASSERT(local_coordinator);
-            ThreadSafeReference active_after = local_coordinator->get_unbound_realm();
-            SharedRealm frozen_before = local_coordinator->get_realm(config, previous_version);
+    // The conditions here are asymmetric because if we have *either* a before
+    // or after callback we need to make sure to initialize the local schema
+    // before the client reset happens.
```

Comment:
I had to think hard about whether this change was okay. There are now different code paths when making a local schema change during a async-open/client-reset scenario. Previously, the Realm was always initialized with the user schema, but now it is only initialized if either callback is set by the user. However, whether or not the schema changes are accepted or rejected does not change if they are applied pre or post client reset. The only difference I can see is is slightly different error messaging.

Reply:
I am pretty sure that this is a place where both behaviors have to be okay or other things are broken anyway. The no-callback behavior should be equivalent to opening the Realm with the old schema, getting and processing a client reset, then restarting and opening with the new schema, which is obviously something that should work. If the callback-set behavior ever has a different end state (other than errors being reported in different ways), that's probably a problem?

Reply:
Agreed. Either way should have a equivalent end result. 👍